MekHQ/src/mekhq/campaign/personnel/education/EducationController.java

```diff
@@ -0,0 +1,1879 @@
+
+            return;
+        }
+
+        if (graduationRoll < 99) {
+            campaign.addReport(person.getHyperlinkedName() + ' ' + resources.getString("graduatedTrialPassed.text")
+                    .replaceAll("0", resources.getString("graduatedEasily.text"))
+                    .replaceAll("1", resources.getString("graduatedScientist.text")));
+
+            ServiceLogger.eduClanPassed(person, campaign.getLocalDate(),
+                    resources.getString("graduatedEasily.text"),
+                    resources.getString("graduatedScientist.text"));
+
+            processGraduation(campaign, person, academy, 1, resources);
+
+            return;
+        }
+
+        campaign.addReport(person.getHyperlinkedName() + ' ' + resources.getString("graduatedTrialPassed.text")
+                .replaceAll("0", resources.getString("graduatedEffortlessly.text"))
+                .replaceAll("1", resources.getString("graduatedScientist.text")));
+
+        ServiceLogger.eduClanPassed(person, campaign.getLocalDate(),
+                resources.getString("graduatedEffortlessly.text"),
+                resources.getString("graduatedScientist.text"));
+
+        processGraduation(campaign, person, academy, 2, resources);
+    }
+
+    /**
+     * Graduates a person from an academy and potentially applies a skill bonus.
+     *
+     * @param campaign      the campaign associated with the person's education
+     * @param person        the person who is graduating with a bonus
+     * @param academy       the academy from which the person is graduating
+     * @param bonusCount    the number of bonuses to be added
+     * @param resources     the resource bundle for retrieving localized messages
+     */
+    private static void processGraduation(Campaign campaign, Person person, Academy academy, Integer bonusCount, ResourceBundle resources) {
+        improveSkills(campaign, person, academy, true);
+
+        if ((campaign.getCampaignOptions().isEnableBonuses()) && (bonusCount > 0)) {
+            addBonus(campaign, person, academy, bonusCount, resources);
+        }
+
+        int educationLevel = academy.getEducationLevel(person);
+
+        if (person.getEduHighestEducation() < educationLevel) {
+            person.setEduHighestEducation(educationLevel);
+        }
+    }
+
+    /**
+     * Improves the skills of a given person based on the curriculum of the course they are attending.
+     *
+     * @param person The Person whose skills are being improved.
+     * @param academy The academy the person is attending.
+     * @param isGraduating A boolean value indicating whether the person is graduating from the academy or not.
+     */
+    private static void improveSkills(Campaign campaign, Person person, Academy academy, Boolean isGraduating) {
+        String[] curriculum = academy.getCurriculums().get(person.getEduCourseIndex())
+                .replaceAll(", ", ",").split(",");
```

Comment:
Might I suggest doing the split first the streaming the array and striping the text for blank text on either side.